### DISCUSSION TITLE
How to make AI turn smoothly?

### DISCUSSION THREAD

**Discussion by Jus2Cat** [Sixze/ALS-Refactored#127] [Q&A] [answered]
Hello. AI rotation is too fast in Aiming mode. What is the right way for decrease turn rate? (and make it smooth). I can't do it because I haven't access to edit rotation rate property and I also can't override FaceRotation()
![turn rate](https://user-images.githubusercontent.com/48173145/181104086-d19d60c5-2967-4dce-8f1a-fdcb32585d27.gif)

**Comment by Jus2Cat** [comment, marked as answer]
Hello. If anyone has same problem. Solution is override `AIController::UpdateControlRotation()` function. So I just added RInterp between last and new control rotation like in example below 


`const float YourRotationRate = 7.0f;`

`
NewControlRotation = FMath::RInterpTo(GetControlRotation(), NewControlRotation = (FocalPoint - MyPawn->GetPawnViewLocation()).Rotation(), DeltaTime, YourRotationRate);
`

instead 

`NewControlRotation = (FocalPoint - MyPawn->GetPawnViewLocation()).Rotation();  `

**Reply by JoeGitcpp** [reply]
`const float YourRotationRate = 25.0f;
	APawn* const MyPawn = GetPawn();
	if (MyPawn)
	{
		FRotator NewControlRotation = GetControlRotation();

		// Look toward focus
		const FVector FocalPoint = GetFocalPoint();
		if (FAISystem::IsValidLocation(FocalPoint))
		{
			//NewControlRotation = (FocalPoint - MyPawn->GetPawnViewLocation()).Rotation();
			FRotator TargetRotation = (FocalPoint - MyPawn->GetPawnViewLocation()).Rotation();
			NewControlRotation = FMath::RInterpTo(NewControlRotation, TargetRotation, DeltaTime, YourRotationRate);
		}
		else if (bSetControlRotationFromPawnOrientation)
		{
			NewControlRotation = MyPawn->GetActorRotation();
		}

		// Don't pitch view unless looking at another pawn
		if (NewControlRotation.Pitch != 0 && Cast<APawn>(GetFocusActor()) == nullptr)
		{
			NewControlRotation.Pitch = 0.f;
		}

		SetControlRotation(NewControlRotation);

		if (bUpdatePawn)
		{
			const FRotator CurrentPawnRotation = MyPawn->GetActorRotation();

			if (CurrentPawnRotation.Equals(NewControlRotation, 1e-3f) == false)
			{
				MyPawn->FaceRotation(NewControlRotation, DeltaTime);
			}
		}
	}

}`

**Reply by Joel-BB** [reply]
This doesn't seem to work for me in the initial detection, it still snaps when it detects and then smoothly rotates afterwards, any idea why?
